lib/ecto/type.ex

```diff
@@ -875,6 +884,94 @@ defmodule Ecto.Type do
     end
   end
 
+  @doc """
+  Checks if two terms are equal.
+
+  Depending on the given `type` performs a structural or semantical comparison.
+
+  ## Examples
+
+      iex> equal?(:integer, 1, 1)
+      true
+      iex> equal?(:decimal, Decimal.new("1"), Decimal.new("1.00"))
+      true
+
+  """
+  @spec equal?(t, term, term) :: boolean
+  def equal?(type, term1, term2) do
+    if requires_semantic_comparison?(type) do
+      do_equal?(type, term1, term2)
+    else
+      term1 == term2
+    end
+  end
+
+  defp requires_semantic_comparison?(:decimal),
+    do: true
+  defp requires_semantic_comparison?(:date),
+    do: false
+  defp requires_semantic_comparison?(type) when type in @calendar,
+    do: true
+  defp requires_semantic_comparison?(type) when type in @base,
+    do: false
+  defp requires_semantic_comparison?({:array, type}),
+    do: requires_semantic_comparison?(type)
+  defp requires_semantic_comparison?({:map, type}),
+    do: requires_semantic_comparison?(type)
+  defp requires_semantic_comparison?(module) when is_atom(module) do
+    function_exported?(module, :equal?, 2)
```

Comment:
This will return false if the module is not loaded and I think there is no guarantee the module will be loaded by then.